scipy/sparse/linalg/_isolve/iterative.py

```diff
@@ -1,496 +1,636 @@
-
-Returns
--------
-x : ndarray
-    The converged solution.
-info : integer
-    Provides convergence information:
-        0  : successful exit
-        >0 : convergence to tolerance not achieved, number of iterations
-        <0 : illegal input or breakdown
-
-Other Parameters
-----------------
-x0 : ndarray
-    Starting guess for the solution.
-tol, atol : float, optional
-    Tolerances for convergence, ``norm(residual) <= max(tol*norm(b), atol)``.
-    The default for ``atol`` is ``'legacy'``, which emulates
-    a different legacy behavior.
-
-    .. warning::
-
-       The default value for `atol` will be changed in a future release.
-       For future compatibility, specify `atol` explicitly.
-maxiter : integer
-    Maximum number of iterations.  Iteration will stop after maxiter
-    steps even if the specified tolerance has not been achieved.
-M : {sparse matrix, ndarray, LinearOperator}
-    Preconditioner for A.  The preconditioner should approximate the
-    inverse of A.  Effective preconditioning dramatically improves the
-    rate of convergence, which implies that fewer iterations are needed
-    to reach a given error tolerance.
-callback : function
-    User-supplied function to call after each iteration.  It is called
-    as callback(xk), where xk is the current solution vector.
-"""
-
-
-def _stoptest(residual, atol):
-    """
-    Successful termination condition for the solvers.
-    """
-    resid = np.linalg.norm(residual)
-    if resid <= atol:
-        return resid, 1
-    else:
-        return resid, 0
+from scipy.linalg import get_blas_funcs, get_lapack_funcs
 
+__all__ = ['bicg', 'bicgstab', 'cg', 'cgs', 'gmres', 'qmr']
 
-def _get_atol(tol, atol, bnrm2, get_residual, routine_name):
-    """
-    Parse arguments for absolute tolerance in termination condition.
+
+def bicg(A, b, x0=None, tol=None, maxiter=None, M=None, callback=None,
+         atol=0., rtol=1e-5):
+    """Use BIConjugate Gradient iteration to solve ``Ax = b``.
 
     Parameters
     ----------
-    tol, atol : object
-        The arguments passed into the solver routine by user.
-    bnrm2 : float
-        2-norm of the rhs vector.
-    get_residual : callable
-        Callable ``get_residual()`` that returns the initial value of
-        the residual.
-    routine_name : str
-        Name of the routine.
+    A : {sparse matrix, ndarray, LinearOperator}
+        The real or complex N-by-N matrix of the linear system.
+        Alternatively, ``A`` can be a linear operator which can
+        produce ``Ax`` and ``A^T x`` using, e.g.,
+        ``scipy.sparse.linalg.LinearOperator``.
+    b : ndarray
+        Right hand side of the linear system. Has shape (N,) or (N,1).
+    x0 : ndarray
+        Starting guess for the solution.
+    rtol, atol : float, optional
+        Parameters for the convergence test. For convergence,
+        ``norm(b - A @ x) <= max(tol*norm(b), atol)`` should be satisfied.
```

Comment:
should probably use the non-deprecated argument, same in other places
```suggestion
        ``norm(b - A @ x) <= max(rtol*norm(b), atol)`` should be satisfied.
```

Reply:
🤦 Thanks for catching that